### PR TITLE
Add menu item to Test Explorer context menu allowing Ignore/Unignore of unit test

### LINKED ISSUE #2964: Unit Test>Test Explorer>Right Click "Disable/Enable Test"
For RD unit testing I know you can put an annotation in to skip over a test. That has to be done manually AFAIK right row. A Right Click option would let you add/remove the annotation easily.

When a test has been disabled the background color could be a grey color or an additional could field of "Disabled" could be populated with an X or some other indicator. Just noting down what comes to mind.

### PR DESCRIPTION
**Pull request by IvenBach**
Close #2964.

TODO:
- Create discrete separate command for Ignore/Unignore that will add/remove the annotation respectively. Following this train of thought there shouldn't be any sharing for properties, one for each (IE `ExecuteToggleTestAnnotationCommand` needs some love).
- `IsSelectionATest` is a lie. Make it check what it's claiming.

Question: 
- Is how I've done the visibility within the XAMl for `TestExplorerControl` appropriate? I thought about having it be a property on the VM but that didn't feel right when I was trying to code it up.

### CODE REVIEW

File: Rubberduck.Core/UI/Converters/BoolToVisibilityConverter.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Globalization;
+using System.Windows;
+using System.Windows.Data;
+
+namespace Rubberduck.UI.Converters
+{
+    class BoolToVisibilityConverter : IValueConverter
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/Converters/BoolToVisibilityConverter.cs]: This class is already built into the framework. 

https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.booleantovisibilityconverter?view=netframework-4.8

File: Rubberduck.Core/UI/UnitTesting/TestExplorerControl.xaml [hunk 2]
```diff
@@ -54,9 +54,14 @@
...
             <BitmapImage x:Key="OutcomeSucceeded" UriSource="pack://application:,,,/Rubberduck.Resources;component/Icons/Fugue/tick-circle.png" />
 
+            <BitmapImage x:Key="IgnoreTestImage" UriSource="pack://application:,,,/Rubberduck.Resources;component/Icons/Fugue/flask-empty.png" />
+            <BitmapImage x:Key="UnignoreTestImage" UriSource="pack://application:,,,/Rubberduck.Resources;component/Icons/Fugue/flask-undo.png" />
+
             <local:TestOutcomeImageSourceConverter x:Key="OutcomeIconConverter" />
             <converters:MillisecondToTimeMagnitudeConverter x:Key="FormattedTime" />
             <converters:InvertBoolValueConverter x:Key="InvertBoolValue" />
+            <converters:BoolToVisibilityConverter x:Key="BoolToVisibility" />
+            <converters:BoolToHiddenVisibilityConverter x:Key="BoolToHiddenVisibility" />
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerControl.xaml]: It doesn’t look like `BoolToHidden` converter is being used.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 3]
```diff
@@ -100,6 +108,9 @@ public TestMethodViewModel MouseOverTest
                 }
                 _mouseOverTest = value;
                 OnPropertyChanged();
+                //OnPropertyChanged(nameof(TestToggleLabel));
+                //OnPropertyChanged(nameof(IsTestSelected));
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Please remove commented out code.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 4]
```diff
@@ -215,6 +234,32 @@ private void HandleTestCompletion(object sender, TestCompletedEventArgs e)
...
+            }
+        }
+
+        public string UnignoreTestLabel => RubberduckUI.TestExplorer_TestToggle_Unignore;
+        public string IgnoreTestLabel => RubberduckUI.TestExplorer_TestToggle_Ignore;
+
+        public bool IsSelectionATest => SelectedItem != null;
+        private TestMethod _selectedTestMethod => ((TestMethodViewModel)SelectedItem).Method;
+
+        public bool DisplayUnignoreTestLabel => IsSelectionATest && _selectedTestMethod.IsIgnored;
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Could be implemented as the inverse of the other property. 

`public bool DisplayUnignoreTestLabel => !DisplayIgnoreTesrLabel`

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 5]
```diff
@@ -341,6 +388,28 @@ private void ExecuteResetResultsCommand(object parameter)
...
+        {
+            var selectedTestMethodViewModel = (TestMethodViewModel)SelectedItem;
+            var testMethod = selectedTestMethodViewModel.Method;
+
+            var annotationUpdater = new AnnotationUpdater();
+            var rewriteSession = RewritingManager.CheckOutCodePaneSession();
+
+            if (testMethod.IsIgnored)
+            {
+                var ignoreTestAnnotation = testMethod.Declaration.Annotations.First(iannotation => iannotation.AnnotationType == Parsing.Annotations.AnnotationType.IgnoreTest);
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Is it possible for `Annotations` to be empty here? If it’s guaranteed to have at least one element, this is fine.

File: Rubberduck.Parsing/QualifiedContext.cs [hunk 6]
```diff
@@ -39,19 +39,19 @@ public override bool Equals(object obj)
 
         public static bool operator ==(QualifiedContext context1, QualifiedContext context2)
         {
-            if ((object)context1 == null)
+            if (context1 is null)
             {
-                return (object)context2 == null;
+                return context2 is null;
             }
```
Comment (rubberduck203) [inline Rubberduck.Parsing/QualifiedContext.cs]: ++

File: Rubberduck.Parsing/Rewriter/RewriterInfo/RewriterInfo.cs [hunk 7]
```diff
@@ -5,25 +5,21 @@ namespace Rubberduck.Parsing.Rewriter.RewriterInfo
...
 
-        public string Replacement { get { return _replacement; } }
-        public int StartTokenIndex { get { return _startTokenIndex; } }
-        public int StopTokenIndex { get { return _stopTokenIndex; } }
+        public string Replacement { get; }
+        public int StartTokenIndex { get; }
+        public int StopTokenIndex { get; }
 
-        public static RewriterInfo None { get { return default(RewriterInfo); } }
+        public static RewriterInfo None => default;
```
Comment (rubberduck203) [inline Rubberduck.Parsing/Rewriter/RewriterInfo/RewriterInfo.cs]: +100. Nice cleanup.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 8]
```diff
@@ -215,6 +234,32 @@ private void HandleTestCompletion(object sender, TestCompletedEventArgs e)
             Tests.Refresh();
         }
 
+        public IRewritingManager RewritingManager { get; }
+
+        public string TestToggleLabel
```
Comment (rubberduck203) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: This property appears to be unused?

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 4 above)
Comment (IvenBach) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: I originally tried to set it up this way. Unfortunately it wasn't accurate and I took a while to figure out. 

This is a super fiddly bit to explain so bear with me. Below is the TestExplorer and anything within the red border will show a valid context menu. If you leave that area then [SelectedItem](https://github.com/IvenBach/Rubberduck/blob/Issue2964_TestExplorerDisableTestViaContextMenu/Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs#L97) will be set to null. 
IsCursorOverTest

As suggested a direct inversion of `DisplayIgnoreTestLabel` would have the `MenuItem` display when it isn't applicable, displayed below in the left-hand gutter. While it may appear as though the `MenuItem` is applicable and would visibly look like it'd apply to the last test the mouse was over `Selection` is `null` at this point and an error is thrown and caught.
Direct inversion shows nonapplicable MenuItem

For that reason the properties are written in that specific way. This does lead into an edge case I found. When you display the context menu, and while the menu is still displayed, move to another test and use `Esc` do dismiss the menu. Without moving the mouse bring up the menu again. Because the mouse wasn't moved `Selection` is still referencing the initially moused over test. Is this worth spending time fretting over?

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 5 above)
Comment (IvenBach) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: While it is possible I don't believe it will/should be. I'm working under the assumption that `@TestMethod` will always there. With the ability to ignore a test I can't envision a reason as to why it would need to be removed.

File: Rubberduck.Parsing/Rewriter/RewriterInfo/RewriterInfo.cs (hunk 7 above)
Comment (IvenBach) [reply Rubberduck.Parsing/Rewriter/RewriterInfo/RewriterInfo.cs]: Just following along with what's pointed out to me by VS.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 5 above)
Comment (retailcoder) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: The assumption that a `TestMethod` will always have a `@TestMethod` annotation is fundamentally correct; in 1.0 the annotation was optional and we could discover tests by name (e.g. any public parameterless method in a standard module with a name that starts (ends?) with "Test", is a test), I don't think we will be rolling that back.

That said the use of `.First` instead of `.Single` is legit (there *could* be illegal duplicate annotations of that type), but IMO it warrants a comment e.g. `// ignore illegal duplicate annotations`.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 4 above)
Comment (retailcoder) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: How about `public bool DisplayUnignoreTestLabel => Selection != null && !DisplayIgnoreTesrLabel`?

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 8 above)
Comment (retailcoder) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Should be a private field, unless it's accessed from outside the class... or removed altogether if it's just unused.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 9]
```diff
@@ -215,6 +231,15 @@ private void HandleTestCompletion(object sender, TestCompletedEventArgs e)
             Tests.Refresh();
         }
 
+        public IRewritingManager RewritingManager { get; }
+
+        public string UnignoreTestLabel => RubberduckUI.TestExplorer_TestToggle_Unignore;
+        public string IgnoreTestLabel => RubberduckUI.TestExplorer_TestToggle_Ignore;
```
Comment (Vogel612) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: AFAICT these should be retrieved in the XAML and not exposed in the ViewModel. At least the other MenuItems in the TestExplorer are using `{Resx [...]}` instead of `{Binding [...]}`

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 10]
```diff
@@ -215,6 +231,15 @@ private void HandleTestCompletion(object sender, TestCompletedEventArgs e)
...
+
+        private TestMethod _mousedOverTestMethod => ((TestMethodViewModel)SelectedItem).Method;
```
Comment (Vogel612) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: if we check for `SelectedItem` being null, this shouldn't assume it is not. The reason for that is that we can't ensure this is only evaluated in `Display(Uni|I)gnoreTestLabel`. It might be easier to have something like `SelectedItemIgnoreMatches(bool ignoreState) => (((TestMethodViewModel)SelectedItem)?.Method.IsIgnored ?? !ignoreState) == ignoreState;`. This is a bit... dense, though, so a proper method would probably be best ;)

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 11]
```diff
@@ -341,6 +369,28 @@ private void ExecuteResetResultsCommand(object parameter)
...
+            var annotationUpdater = new AnnotationUpdater();
+            annotationUpdater.AddAnnotation(rewriteSession, qualifiedContext, Parsing.Annotations.AnnotationType.IgnoreTest);
+
+            rewriteSession.TryRewrite();
+        }
+
+        private void ExecuteUnignoreTestCommand(object parameter)
+        {
+            var rewriteSession = RewritingManager.CheckOutCodePaneSession();
+            var ignoreTestAnnotation = _mousedOverTestMethod.Declaration.Annotations.First(iannotation => iannotation.AnnotationType == Parsing.Annotations.AnnotationType.IgnoreTest);
```
Comment (Vogel612) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: would it not be necessary to mark "all" the annotations of that type for removal?

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 12]
```diff
@@ -341,6 +369,28 @@ private void ExecuteResetResultsCommand(object parameter)
...
+
+            var annotationUpdater = new AnnotationUpdater();
```
Comment (MDoerner) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Dito

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs [hunk 13]
```diff
@@ -341,6 +369,28 @@ private void ExecuteResetResultsCommand(object parameter)
             Tests.Refresh();
         }
 
+        private void ExecuteIgnoreTestCommand(object parameter)
+        {
+            var rewriteSession = RewritingManager.CheckOutCodePaneSession();
+            var qualifiedContext = new QualifiedContext(_mousedOverTestMethod.Declaration.QualifiedModuleName, _mousedOverTestMethod.Declaration.Context);
```
Comment (MDoerner) [inline Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: Why not use the overload of `AddAnnotation` taking a declaration?

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 11 above)
Comment (MDoerner) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: I would also say that removing all seems to be more sensible. This will just require replacing `First` with `Where` and adding an `s` to the variable name.

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 13 above)
Comment (IvenBach) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: I derped when I was coding and didn't see it...

File: Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs (hunk 10 above)
Comment (IvenBach) [reply Rubberduck.Core/UI/UnitTesting/TestExplorerViewModel.cs]: I'll need more breadcrumbs to follow and understand this.

File: RubberduckTests/UnitTesting/MockedTestExplorer.cs [hunk 14]
```diff
@@ -14,7 +14,7 @@ public MockedTestExplorer(MockedTestExplorerModel model)
             Vbe = model.Engine.Vbe.Object;
             State = model.Engine.ParserState;
             Model = model.Model;
-            ViewModel = new TestExplorerViewModel(null, Model, ClipboardWriter.Object, null, null);
+            ViewModel = new TestExplorerViewModel(null, Model, ClipboardWriter.Object, null, null, null, null);
```
Comment (bclothier) [inline RubberduckTests/UnitTesting/MockedTestExplorer.cs]: As a general rule, we should avoid newing things in tests but rather reference to a single static method `Arranage...`. That helps us avoid having a ugly refactoring session when we need to modify the parameters for the ctor. This puppy is already up to 7 parameters 😨 so it's surely itching to be refactored.

**Comment by IvenBach** [comment]
https://github.com/IvenBach/Rubberduck/blob/Issue2964_TestExplorerDisableTestViaContextMenu/Rubberduck.Core/UI/UnitTesting/TestExplorerControl.xaml#L421 for not-so-distant-future-Iven to attempt to add `(Uni|I)gnoreSelectedTestsCommand`.
